bpo-35823: subprocess: Use vfork() instead of fork() on Linux when safe

izbyshev opened the pull request:
When used to run a new executable image, fork() is not a good choice
for process creation, especially if the parent has a large working set:
fork() needs to copy page tables, which is slow, and may fail on systems
where overcommit is disabled, despite that the child is not going to
touch most of its address space.

Currently, subprocess is capable of using posix_spawn() instead, which
normally provides much better performance. However, posix_spawn() does not
support many of child setup operations exposed by subprocess.Popen().
Most notably, it's not possible to express `close_fds=True`, which
happens to be the default, via posix_spawn(). As a result, most users
can't benefit from faster process creation, at least not without
changing their code.

However, Linux provides vfork() system call, which creates a new process
without copying the address space of the parent, and which is actually
used by C libraries to efficiently implement posix_spawn(). Due to sharing
of the address space and even the stack with the parent, extreme care
is required to use vfork(). At least the following restrictions must hold:

* No signal handlers must execute in the child process. Otherwise, they
  might clobber memory shared with the parent, potentially confusing it.

* Any library function called after vfork() in the child must be
  async-signal-safe (as for fork()), but it must also not interact with any
  library state in a way that might break due to address space sharing
  and/or lack of any preparations performed by libraries on normal fork().
  POSIX.1 permits to call only execve() and _exit(), and later revisions
  remove vfork() specification entirely. In practice, however, almost all
  operations needed by subprocess.Popen() can be safely implemented on
  Linux.

* Due to sharing of the stack with the parent, the child must be careful
  not to clobber local variables that are alive across vfork() call.
  Compilers are normally aware of this and take extra care with vfork()
  (and setjmp(), which has a similar problem).

* In case the parent is privileged, special attention must be paid to vfork()
  use, because sharing an address space across different privilege domains
  is insecure[1].

This patch adds support for using vfork() instead of fork() on Linux
when it's possible to do safely given the above. In particular:

* vfork() is not used if credential switch is requested. The reverse case
  (simple subprocess.Popen() but another application thread switches
  credentials concurrently) is not possible for pure-Python apps because
  subprocess.Popen() and functions like os.setuid() are mutually excluded
  via GIL. We might also consider to add a way to opt-out of vfork() (and
  posix_spawn() on platforms where it might be implemented via vfork()) in
  a future PR.

* vfork() is not used if `preexec_fn != None`.

With this change, subprocess will still use posix_spawn() if possible, but
will fallback to vfork() on Linux in most cases, and, failing that,
to fork().

[1] https://ewontfix.com/7

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-35823](https://bugs.python.org/issue35823) -->
https://bugs.python.org/issue35823
<!-- /issue-number -->

izbyshev commented on Modules/_posixsubprocess.c:
```diff
@@ -30,6 +30,12 @@
+#if defined(__linux__) && defined(HAVE_SIGNAL_H) && \
+    defined(HAVE_PTHREAD_SIGMASK) && !defined(HAVE_BROKEN_PTHREAD_SIGMASK)
+# include <signal.h>
+# define VFORK_USABLE 1
```
We may need to add a configure test for `vfork` if possibility of Linux without `vfork` is considered worthy.

gpshead replied:
Based on past experience with my own assumptions about availability of API calls.. yes, we should go ahead with a configure.ac check.

izbyshev commented on Modules/_posixsubprocess.c:
```diff
@@ -383,6 +389,43 @@ _close_open_fds_maybe_unsafe(long start_fd, PyObject* py_fds_to_keep)
+         * internally (e.g. for thread cancellation), so simply
+         * skip errors here. */
+        if (sigaction(sig, NULL, &sa) == -1) {
+            continue;
```
Here we have a problem that we need about `_NSIG` (usually 64 or 128) syscalls  to reset signal handlers. Glibc does exactly the same in `posix_spawn`, so no regression compared to glibc, but `musl` implements a kind of caching which avoids extra syscalls but is impossible to implement outside of a C library (without wrapping standard C functions).

One way to optimize this is to parse `SigCgt` field from `/proc/self/status`.

ronaldoussoren replied:
I wouldn't bother optimising unless there is a clear performance reason to do so. Especially because glibc also resets all signal handlers.

gpshead replied:
Agreed.  I also question whether parsing `proc/self/status` for `SigCgt` would even be faster (that'll depend on platform and NSIG).  It'd certainly be more complex which is something I prefer to avoid when possible.

izbyshev replied:
I agree, I also don't want to optimize in the initial version, at least not with `/proc/self/status`. I've also discovered that `SigCgt` is incorrect (always zero) in Windows Subsystem for Linux (tested on Windows 10 1803).

gpshead reviewed (commented):
overall I think this PR is looking good.

izbyshev commented on Modules/_posixsubprocess.c:
```diff
@@ -405,6 +451,9 @@ child_exec(char *const exec_array[],
            int call_setsid,
+#ifdef VFORK_USABLE
+           const sigset_t *old_sigs,
+#endif
```
To avoid `#ifdef` here and in `do_fork_exec()` while still not depending on `sigset_t` unconditionally, I can change the type of this `old_sigs` to `void *` and add a cast when it's actually used. I'm not sure what's worse -- ugly `#ifdef`s or a bit hacky parameter.

Jongy replied:
I tried removing `_Py_NO_INLINE` and I didn't get any `-Wclobbered` (with GCC 9.3.0). But anyway - even if we get the warning with other GCC versions - we can make it `_Py_NO_INLINE` only if `VFORK_USABLE`. Then, we can remove the parameter `#ifdef` here and in `do_fork_exec`, pass `NULL` in the non-`VFORK_USABLE` case, and trust the compiler to inline it all and eliminate the unneeded parameter since both functions are inlinable.

izbyshev replied:
Unfortunately, GCC doesn't inline `child_exec()` even I try to force it with `always_inline`.

But those `#ifdef`s in parameter/argument lists seem much uglier to me than a `void *` parameter, and overhead of passing a useless argument in non-`VFORK_USABLE` case should be negligible, so I'm leaning to getting rid of `#ifdef`s here for the next patch version.

Commit:
bpo-35823: subprocess: Use vfork() instead of fork() on Linux when safe

When used to run a new executable image, fork() is not a good choice
for process creation, especially if the parent has a large working set:
fork() needs to copy page tables, which is slow, and may fail on systems
where overcommit is disabled, despite that the child is not going to
touch most of its address space.

Currently, subprocess is capable of using posix_spawn() instead, which
normally provides much better performance. However, posix_spawn() does not
support many of child setup operations exposed by subprocess.Popen().
Most notably, it's not possible to express `close_fds=True`, which
happens to be the default, via posix_spawn(). As a result, most users
can't benefit from faster process creation, at least not without
changing their code.

However, Linux provides vfork() system call, which creates a new process
without copying the address space of the parent, and which is actually
used by C libraries to efficiently implement posix_spawn(). Due to sharing
of the address space and even the stack with the parent, extreme care
is required to use vfork(). At least the following restrictions must hold:

* No signal handlers must execute in the child process. Otherwise, they
  might clobber memory shared with the parent, potentially confusing it.

* Any library function called after vfork() in the child must be
  async-signal-safe (as for fork()), but it must also not interact with any
  library state in a way that might break due to address space sharing
  and/or lack of any preparations performed by libraries on normal fork().
  POSIX.1 permits to call only execve() and _exit(), and later revisions
  remove vfork() specification entirely. In practice, however, almost all
  operations needed by subprocess.Popen() can be safely implemented on
  Linux.

* Due to sharing of the stack with the parent, the child must be careful
  not to clobber local variables that are alive across vfork() call.
  Compilers are normally aware of this and take extra care with vfork()
  (and setjmp(), which has a similar problem).

* In case the parent is privileged, special attention must be paid to vfork()
  use, because sharing an address space across different privilege domains
  is insecure[1].

This patch adds support for using vfork() instead of fork() on Linux
when it's possible to do safely given the above. In particular:

* vfork() is not used if credential switch is requested. The reverse case
  (simple subprocess.Popen() but another application thread switches
  credentials concurrently) is not possible for pure-Python apps because
  subprocess.Popen() and functions like os.setuid() are mutually excluded
  via GIL. We might also consider to add a way to opt-out of vfork() (and
  posix_spawn() on platforms where it might be implemented via vfork()) in
  a future PR.

* vfork() is not used if `preexec_fn != None`.

With this change, subprocess will still use posix_spawn() if possible, but
will fallback to vfork() on Linux in most cases, and, failing that,
to fork().

[1] https://ewontfix.com/7

izbyshev commented on Modules/_posixsubprocess.c:
```diff
@@ -835,39 +980,55 @@ subprocess_fork_exec(PyObject* self, PyObject *args)
+         * such as QEMU user-mode emulation, the parent won't be blocked,
+         * but it won't share the address space with the child,
+         * so it's still safe to unblock the signals. */
+        pthread_sigmask(SIG_SETMASK, old_sigmask, NULL);
```
I don't check `pthread_sigmask()` return value here and elsewhere. It's not supposed to fail if called with sane arguments. But if we do want to report the failure to the caller here, we need to kill and reap the child process, and even then the caller will be left with the changed signal mask. Logically, we're leaving a "critical section" here, so attempting to handle failures looks like an overkill to me.
That said, we can sanely handle a hypothetical failure of the first `pthread_sigmask()` call (before `do_fork_exec()`) if desired.

gpshead replied:
Agreed, it does not seem reasonable for `pthread_sigmask()` to return an error in this situation.

gpshead committed:
Prevent vfork() when call_setsid is set.

It isn't clear if setsid() is safe after vfork(); better safe than sorry.

Also adds a couple comments and some debug build assertions.

gpshead commented on Modules/_posixsubprocess.c:
```diff
@@ -835,39 +980,55 @@ subprocess_fork_exec(PyObject* self, PyObject *args)
-         * http://www.opengroup.org/onlinepubs/009695399/functions/xsh_chap02_04.html.
+    const void *old_sigmask = NULL;
+#ifdef VFORK_USABLE
+    /* Use vfork() only if it's safe. See the comment above child_exec(). */
```
I added !call_setsid to this list.

gpshead commented on Modules/_posixsubprocess.c:
```diff
@@ -506,6 +576,13 @@ child_exec(char *const exec_array[],
+#ifdef VFORK_USABLE
+    if (child_sigmask) {
+        reset_signal_handlers(child_sigmask);
+        pthread_sigmask(SIG_SETMASK, child_sigmask, NULL);
```
I added an error check on this one because it is doable in this spot even though unlikely.
I also verified that reset_signal_handlers has no need to report any errors.

gpshead reviewed (approved):
Thanks for this work!  Waiting on CI and doing some test runs myself.  I expect to merge it shortly if that goes well.
